src/openvic-simulation/economy/Producer.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <map>
+
+#include "openvic-simulation/economy/Good.hpp"
+#include "openvic-simulation/economy/ProductionType.hpp"
+#include "openvic-simulation/types/fixed_point/FixedPoint.hpp"
+#include "openvic-simulation/utility/Getters.hpp"
+
+using money_t = OpenVic::fixed_point_t;
+using good_quantity_t = OpenVic::fixed_point_t;
+
+namespace OpenVic {
+	class Producer {
+	public:
+		ProductionType PROPERTY(production_type)
+	};
+
+	class Manufacturer : Producer {
+	public:
+		Good::good_map_t PROPERTY(stockpile);
+	};
+
+	class ArtisanalProducer final : Manufacturer {
+	public:
+		good_quantity_t PROPERTY(current_production);
+		Good::good_map_t PROPERTY(current_needs);
+	};
+
+	class NonArtisanalProducer : Producer {
+	private:
+		Pop::pop_size_t get_max_workforce_size() {
+			return static_cast<Pop::pop_size_t>(
+				static_cast<fixed_point_t>(production_type.get_base_workforce_size()) * size_multiplier
+			);
+		}
+
+	public:
+		money_t PROPERTY(revenue_yesterday);
+		good_quantity_t PROPERTY(output_quantity_yesterday);
+		good_quantity_t PROPERTY(unsold_quantity_yesterday);
+		fixed_point_t PROPERTY(size_multiplier);
+		std::map<Pop const*, Pop::pop_size_t> PROPERTY(employees)
+	};
+
+	class FactoryProducer final : NonArtisanalProducer, Manufacturer {
```

Comment:
- When extending a class, by default everything is made private, even if it's public in the base class. To get around this, you either need to change the `class`es to `struct`s or you need to change them here to `: public NonArtisanalProducer, public Manufacturer`.
- As `NonArtisanalProducer` and `Manufacturer` both extend `Producer`, by default `FactoryProducer` will contain two separate copies of `Producer`. To get around this you need to use virtual inheritance, so `: virtual NonArtisanalProducer, virtual Manufacturer` (and make them `public`, unless you change them to `struct`s). You'll also need to add a virtual destructor to `Producer`. I'd recommend you look up virtual inheritance, as it's a bit of a weird concept in C++.

Reply:
I did not expect that. I read some stuff about c++ not having interfaces (which I'd love to use here).
I'll make them public & virtual.